pubsub/rabbitmq/rabbitmq.go

```diff
@@ -409,32 +432,49 @@ func (r *rabbitMQ) subscribeForever(req pubsub.SubscribeRequest, queueName strin
 			return
 		}
 
-		// print the error if the subscriber is running.
-		if err != nil {
-			r.logger.Errorf("%s error in subscriber for %s in %s: %v", logMessagePrefix, queueName, errFuncName, err)
+		// print errors if the subscriber is running.
+		if len(errs) != 0 {
+			for i := range errs {
+				r.logger.Errorf("%s error in subscriber for %s in %s: %v", logMessagePrefix, queueName, errFuncName, errs[i])
+			}
 		}
 
-		if mustReconnect(channel, err) {
+		if mustReconnect(channel, errs) {
 			r.logger.Warnf("%s subscriber is reconnecting in %s ...", logMessagePrefix, r.metadata.reconnectWait.String())
 			time.Sleep(r.metadata.reconnectWait)
 			r.reconnect(connectionCount)
 		}
 	}
 }
 
-func (r *rabbitMQ) listenMessages(channel rabbitMQChannelBroker, msgs <-chan amqp.Delivery, topic string, handler pubsub.Handler) error {
-	var err error
-	for d := range msgs {
-		switch r.metadata.concurrency {
-		case pubsub.Single:
-			err = r.handleMessage(channel, d, topic, handler)
-		case pubsub.Parallel:
-			go func(channel rabbitMQChannelBroker, d amqp.Delivery, topic string, handler pubsub.Handler) {
-				err = r.handleMessage(channel, d, topic, handler)
-			}(channel, d, topic, handler)
+func (r *rabbitMQ) listenMessages(channel rabbitMQChannelBroker, msgs <-chan amqp.Delivery, topic string, handler pubsub.Handler) []error {
+	switch r.metadata.concurrency {
+	case pubsub.Single:
+		for d := range msgs {
+			if e := r.handleMessage(channel, d, topic, handler); e != nil {
+				if mustReconnect(channel, []error{e}) {
+					return []error{e}
+				}
+			}
 		}
-		if (err != nil) && mustReconnect(channel, err) {
-			return err
+	case pubsub.Parallel:
+		ec := ErrorCollection{errors: []error{}}
+		stopCh := make(chan struct{})
+		var wg sync.WaitGroup
+		for {
+			select {
+			case <-stopCh:
+				wg.Wait()
+				return ec.GetErrors()
+			case d := <-msgs:
+				wg.Add(1)
+				go func(channel rabbitMQChannelBroker, d amqp.Delivery, topic string, handler pubsub.Handler) {
+					if e := r.handleMessage(channel, d, topic, handler); e != nil && mustReconnect(channel, []error{e}) {
+						ec.Append(e, stopCh)
+					}
+					wg.Done()
+				}(channel, d, topic, handler)
+			}
```

Comment:
Also, I'd more like make `stopCh` as a field of `ErrorCollection` to make the code cleaner. And the param of `Append` can be an error only.

```
case pubsub.Parallel:
		ec := ErrorCollection{errNotify: make(chan struct{})}
		var wg sync.WaitGroup
		for {
			select {
			case <-ec.errNotify:
				wg.Wait()
				return ec.GetErrors()
			case d := <-msgs:
				wg.Add(1)
				go func(channel rabbitMQChannelBroker, d amqp.Delivery, topic string, handler pubsub.Handler) {
					if e := r.handleMessage(channel, d, topic, handler); e != nil && mustReconnect(channel, []error{e}) {
						ec.Append(e)
					}
					wg.Done()
				}(channel, d, topic, handler)
			}
		}
	}
```
And for ErrorCollection definition:
```
type ErrorCollection struct {
	errors    []error
	mux       sync.Mutex
	errNotify chan struct{}
}

func (c *ErrorCollection) Append(e error) {
	c.mux.Lock()
	if len(c.errors) == 0 {
		close(c.errNotify)
	}
	c.errors = append(c.errors, e)
	c.mux.Unlock()
}
```

Reply:
@Taction - Done. And I am trying to make the ErrorCollection shared since we may use it in other packages. But I am not sure whether the sharing way is a good practice in the Golang. https://github.com/gdlcf88/components-contrib/commit/2a64a565beae7d90b58e3af045505e7cc857f33b

Reply:
This is a good idea, but since this is a small change and there have no other packages using it for now, I would rather refactor the code when anywhere else uses it.